# TELCO PROJECT

# GOAL:
<div class="alert alert-success">

* Words go Here
* More Words go Here
</div>

# IMPORTS
<div class="alert alert-success">
</div>

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import os

from sklearn.model_selection import train_test_split
import sklearn.preprocessing
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

#import warnings
#warnings.filterwarnings("ignore")

from scipy import stats
import re

import wrangle as w
import QMCBT_explore_evaluate as ee
#import QMCBT_modeling as m

α = 0.05
alpha = 0.05

In [2]:
# Acquire telco_churn data from Codeup SQL database
df = w.get_telco_churn_df()

# EXPLORE
<div class="alert alert-success">
    
    
</div>

In [3]:
df.T

0                    1     \
customer_id                          0002-ORFBO           0003-MKNFE   
payment_type_id                               2                    2   
contract_type_id                              2                    1   
internet_service_type_id                      1                    1   
gender                                   Female                 Male   
senior_citizen                                0                    0   
partner                                     Yes                   No   
dependents                                  Yes                   No   
tenure                                        9                    9   
phone_service                               Yes                  Yes   
multiple_lines                               No                  Yes   
online_security                              No                   No   
online_backup                               Yes                   No   
device_protection                            No                   No   
tech_support                                Yes                   No   
streaming_tv                                Yes                   No   
streaming_movies                             No                  Yes   
paperless_billing                           Yes                   No   
monthly_charges                            65.6                 59.9   
total_charges                             593.3                542.4   
churn                                        No                   No   
internet_service_type                       DSL                  DSL   
contract_type                          One year       Month-to-month   
payment_type                       Mailed check         Mailed check   
churn_month                                 NaN                  NaN   
contract_type_id.1                            2                    1   
paperless_billing.1                         Yes                   No   
gender.1                                 Female                 Male   
senior_citizen.1                              0                    0   
partner.1                                   Yes                   No   
dependents.1                                Yes                   No   
payment_type_id.1                             2                    2   
monthly_charges.1                          65.6                 59.9   
total_charges.1                           593.3                542.4   
signup_date                 2021-04-21 18:07:34  2021-04-21 18:07:34   
phone_service.1                             Yes                  Yes   
multiple_lines.1                             No                  Yes   
internet_service_type_id.1                    1                    1   
online_security.1                            No                   No   
online_backup.1                             Yes                   No   
device_protection.1                          No                   No   
tech_support.1                              Yes                   No   
streaming_tv.1                              Yes                   No   
streaming_movies.1                           No                  Yes   

                                           2                    3     \
customer_id                          0004-TLHLJ           0011-IGKFF   
payment_type_id                               1                    1   
contract_type_id                              1                    1   
internet_service_type_id                      2                    2   
gender                                     Male                 Male   
senior_citizen                                0                    1   
partner                                      No                  Yes   
dependents                                   No                   No   
tenure                                        4                   13   
phone_service                               Yes                  Yes   
multiple_l

In [4]:
ee.explore_toc()

** CUSTOM EXPLORATION FUNCTIONS
explore_tips: PRINT A LIST OF USEFUL FUNCTIONS, METHODS, AND ATTRIBUTES USED FOR EXPLORATION
nunique_column_all(df): PRINT NUNIQUE OF ALL COLUMNS
nunique_column_objects(df): PRINT NUNIQUE OF COLUMNS THAT ARE OBJECTS
nunique_column_qty(df): PRINT NUNIQUE OF COLUMNS THAT ARE *NOT* OBJECTS
numeric_range(df): COMPUTE RANGE FOR ALL NUMERIC VARIABLES


In [5]:
ee.explore_tips()

** USEFUL EXPLORATORY CODE**
DFNAME.head()
DFNAME.shape
DFNAME.shape[0] #read row count
DFNAME.describe().T
DFNAME.columns.to_list()
DFNAME.COLUMNNAME.value_counts(dropna=False)
DFNAME.dtypes
DFNAME.select_dtypes(include='object').columns
DFNAME.select_dtypes(include='float').columns
pd.crosstab(DFNAME.COLUMN-1, DFNAME.COLUMN-2)


In [6]:
# return (Row, Column)
df.shape

(7043, 44)

In [7]:
# Return non-null count and Dtype for each column
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   customer_id                 7043 non-null   object 
 1   payment_type_id             7043 non-null   int64  
 2   contract_type_id            7043 non-null   int64  
 3   internet_service_type_id    7043 non-null   int64  
 4   gender                      7043 non-null   object 
 5   senior_citizen              7043 non-null   int64  
 6   partner                     7043 non-null   object 
 7   dependents                  7043 non-null   object 
 8   tenure                      7043 non-null   int64  
 9   phone_service               7043 non-null   object 
 10  multiple_lines              7043 non-null   object 
 11  online_security             7043 non-null   object 
 12  online_backup               7043 non-null   object 
 13  device_protection           7043 

In [8]:
df.describe().T

count       mean        std    min   25%    50%  \
payment_type_id             7043.0   2.315633   1.148907   1.00   1.0   2.00   
contract_type_id            7043.0   1.690473   0.833755   1.00   1.0   1.00   
internet_service_type_id    7043.0   1.872923   0.737796   1.00   1.0   2.00   
senior_citizen              7043.0   0.162147   0.368612   0.00   0.0   0.00   
tenure                      7043.0  32.371149  24.559481   0.00   9.0  29.00   
monthly_charges             7043.0  64.761692  30.090047  18.25  35.5  70.35   
contract_type_id.1          7043.0   1.690473   0.833755   1.00   1.0   1.00   
senior_citizen.1            7043.0   0.162147   0.368612   0.00   0.0   0.00   
payment_type_id.1           7043.0   2.315633   1.148907   1.00   1.0   2.00   
monthly_charges.1           7043.0  64.761692  30.090047  18.25  35.5  70.35   
internet_service_type_id.1  7043.0   1.872923   0.737796   1.00   1.0   2.00   

                              75%     max  
payment_type_id              3.00    4.00  
contract_type_id             2.00    3.00  
internet_service_type_id     2.00    3.00  
senior_citizen               0.00    1.00  
tenure                      55.00   72.00  
monthly_charges             89.85  118.75  
contract_type_id.1           2.00    3.00  
senior_citizen.1             0.00    1.00  
payment_type_id.1            3.00    4.00  
monthly_charges.1           89.85  118.75  
internet_service_type_id.1   2.00    3.00

In [9]:
ee.nunique_column_objects(df)

customer_id has 7043 unique values.
gender has 2 unique values.
partner has 2 unique values.
dependents has 2 unique values.
phone_service has 2 unique values.
multiple_lines has 3 unique values.
online_security has 3 unique values.
online_backup has 3 unique values.
device_protection has 3 unique values.
tech_support has 3 unique values.
streaming_tv has 3 unique values.
streaming_movies has 3 unique values.
paperless_billing has 2 unique values.
total_charges has 6531 unique values.
churn has 2 unique values.
internet_service_type has 3 unique values.
contract_type has 3 unique values.
payment_type has 4 unique values.
churn_month has 1 unique values.
paperless_billing.1 has 2 unique values.
gender.1 has 2 unique values.
partner.1 has 2 unique values.
dependents.1 has 2 unique values.
total_charges.1 has 6531 unique values.
signup_date has 73 unique values.
phone_service.1 has 2 unique values.
multiple_lines.1 has 3 unique values.
online_security.1 has 3 unique values.
online_backup.

In [10]:
ee.nunique_column_qty(df)

payment_type_id has 4 unique values.
contract_type_id has 3 unique values.
internet_service_type_id has 3 unique values.
senior_citizen has 2 unique values.
tenure has 73 unique values.
monthly_charges has 1585 unique values.
contract_type_id.1 has 3 unique values.
senior_citizen.1 has 2 unique values.
payment_type_id.1 has 4 unique values.
monthly_charges.1 has 1585 unique values.
internet_service_type_id.1 has 3 unique values.


In [11]:
df.multiple_lines.unique()

array(['No', 'Yes', 'No phone service'], dtype=object)

In [12]:
df.online_security.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [13]:
df.online_backup.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [14]:
df.device_protection.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [15]:
df.tech_support.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [16]:
df.streaming_tv.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [17]:
df.streaming_movies.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [18]:
df.internet_service_type.unique()

array(['DSL', 'Fiber optic', 'None'], dtype=object)

In [19]:
df.contract_type.unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [20]:
df.payment_type.unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

# PREPARE
<div class="alert alert-success">
    
    
</div>

### PREPARE To-Do List

<div class="alert alert-info">

* DROP 
    * customer_id
    * payment_type_id
    * contract_type_id
    * internet_service_type_id
    * churn_month
    * paperless_billing.1
    * gender.1 
    * partner.1 
    * dependents.1 
    * total_charges.1 
    * phone_service.1 
    * multiple_lines.1 
    * online_security.1
    * online_backup.1 
    * device_protection.1
    * tech_support.1
    * streaming_tv.1
    * streaming_movies.1
    * contract_type_id.1 
    * senior_citizen.1 
    * payment_type_id.1 
    * monthly_charges.1 
    * internet_service_type_id.1 
    
* .rename
    * NONE

* .replace
    * churn_month has nulls; replace nulls with 'active' DROP (ALL DATES DUPLICATED).
    * multiple_lines has 3 unique values. REPLACE 'No phone service' with 'No'
    * online_security has 3 unique values. REPLACE 'No internet service' with 'No'
    * online_backup has 3 unique values. REPLACE 'No internet service' with 'No'
    * device_protection has 3 unique values. REPLACE 'No internet service' with 'No'
    * tech_support has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_tv has 3 unique values. REPLACE 'No internet service' with 'No'
    * streaming_movies has 3 unique values. REPLACE 'No internet service' with 'No'

* reformat
    * signup_date 7043 non-null   object. REFORMAT OBJECT astype DATETIME
    * total_charges contains empty " " values. REPLACE empty " " values
    * total_charges 7043 non-null   object. REFORMAT OBJECT astype FLOAT

* encode.map
    * gender has 2 unique values.
    * partner has 2 unique values.
    * dependents has 2 unique values.
    * phone_service has 2 unique values.
    * paperless_billing has 2 unique values.
    * churn has 2 unique values.
    * senior_citizen has 2 unique values.
    * multiple_lines will have 2 unique values. REPLACED 'No phone service' with 'No'
    * online_security will have 2 unique values. REPLACED 'No internet service' with 'No'
    * online_backup will have 2 unique values. REPLACED 'No internet service' with 'No'
    * device_protection will have 2 unique values. REPLACED 'No internet service' with 'No'
    * tech_support will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_tv will have 2 unique values. REPLACED 'No internet service' with 'No'
    * streaming_movies will have 2 unique values. REPLACED 'No internet service' with 'No'

* melt
    * NONE

* pivot
    * internet_service_type has 3 unique values.
    * contract_type has 3 unique values.
    * payment_type has 4 unique values.  

* feature engineer
    * NONE

* Impute
    * NONE

</div>

In [22]:
# Drop duplicate columns
df2 = df.T.drop_duplicates().T
df2.T

0                    1     \
customer_id                        0002-ORFBO           0003-MKNFE   
payment_type_id                             2                    2   
contract_type_id                            2                    1   
internet_service_type_id                    1                    1   
gender                                 Female                 Male   
senior_citizen                              0                    0   
partner                                   Yes                   No   
dependents                                Yes                   No   
tenure                                      9                    9   
phone_service                             Yes                  Yes   
multiple_lines                             No                  Yes   
online_security                            No                   No   
online_backup                             Yes                   No   
device_protection                          No                   No   
tech_support                              Yes                   No   
streaming_tv                              Yes                   No   
streaming_movies                           No                  Yes   
paperless_billing                         Yes                   No   
monthly_charges                          65.6                 59.9   
total_charges                           593.3                542.4   
churn                                      No                   No   
internet_service_type                     DSL                  DSL   
contract_type                        One year       Month-to-month   
payment_type                     Mailed check         Mailed check   
churn_month                               NaN                  NaN   
signup_date               2021-04-21 18:07:34  2021-04-21 18:07:34   

                                         2                    3     \
customer_id                        0004-TLHLJ           0011-IGKFF   
payment_type_id                             1                    1   
contract_type_id                            1                    1   
internet_service_type_id                    2                    2   
gender                                   Male                 Male   
senior_citizen                              0                    1   
partner                                    No                  Yes   
dependents                                 No                   No   
tenure                                      4                   13   
phone_service                             Yes                  Yes   
multiple_lines                             No                   No   
online_security                            No                   No   
online_backup                              No                  Yes   
device_protection                         Yes                  Yes   
tech_support                               No                   No   
streaming_tv                               No                  Yes   
streaming_movies                           No                  Yes   
paperless_billing                         Yes                  Yes   
monthly_charges                          73.9                 98.0   
total_charges                          280.85              1237.85   
churn                                     Yes                  Yes   
internet_service_type             Fiber optic          Fiber optic   
contract_type                  Month-to-month       Month-to-month   
payment_type                 Electronic check     Electronic check   
churn_month                        2022-01-31           2022-01-31   
signup_date               2021-09-21 18:07:34  2020-12-21 18:07:34   

                                         4                        5     \
customer_id                        0013-EXCHZ               0013-MHZWF   
payment_type_id                             2                        4   
contract_type_id                     

In [23]:
df.columns

Index(['customer_id', 'payment_type_id', 'contract_type_id',
       'internet_service_type_id', 'gender', 'senior_citizen', 'partner',
       'dependents', 'tenure', 'phone_service', 'multiple_lines',
       'online_security', 'online_backup', 'device_protection', 'tech_support',
       'streaming_tv', 'streaming_movies', 'paperless_billing',
       'monthly_charges', 'total_charges', 'churn', 'internet_service_type',
       'contract_type', 'payment_type', 'churn_month', 'contract_type_id.1',
       'paperless_billing.1', 'gender.1', 'senior_citizen.1', 'partner.1',
       'dependents.1', 'payment_type_id.1', 'monthly_charges.1',
       'total_charges.1', 'signup_date', 'phone_service.1', 'multiple_lines.1',
       'internet_service_type_id.1', 'online_security.1', 'online_backup.1',
       'device_protection.1', 'tech_support.1', 'streaming_tv.1',
       'streaming_movies.1'],
      dtype='object')

In [24]:
df2.columns

Index(['customer_id', 'payment_type_id', 'contract_type_id',
       'internet_service_type_id', 'gender', 'senior_citizen', 'partner',
       'dependents', 'tenure', 'phone_service', 'multiple_lines',
       'online_security', 'online_backup', 'device_protection', 'tech_support',
       'streaming_tv', 'streaming_movies', 'paperless_billing',
       'monthly_charges', 'total_charges', 'churn', 'internet_service_type',
       'contract_type', 'payment_type', 'churn_month', 'signup_date'],
      dtype='object')

# DROP
<div class="alert alert-warning">

* customer_id
* payment_type_id
* contract_type_id
* internet_service_type_id
* churn_month
* paperless_billing.1
* gender.1 
* partner.1 
* dependents.1 
* total_charges.1 
* phone_service.1 
* multiple_lines.1 
* online_security.1
* online_backup.1 
* device_protection.1
* tech_support.1
* streaming_tv.1
* streaming_movies.1
* contract_type_id.1 
* senior_citizen.1 
* payment_type_id.1 
* monthly_charges.1 
* internet_service_type_id.1 
    
</div>

In [21]:
df = df.drop(columns=['customer_id',
             'payment_type_id',
             'contract_type_id',
             'internet_service_type_id',
             'churn_month',
             'paperless_billing.1',
             'gender.1', 
             'partner.1', 
             'dependents.1', 
             'total_charges.1', 
             'phone_service.1', 
             'multiple_lines.1', 
             'online_security.1',
             'online_backup.1', 
             'device_protection.1',
             'tech_support.1',
             'streaming_tv.1',
             'streaming_movies.1',
             'contract_type_id.1', 
             'senior_citizen.1', 
             'payment_type_id.1', 
             'monthly_charges.1', 
             'internet_service_type_id.1'
            ])

In [22]:
df.T

0                    1     \
gender                              Female                 Male   
senior_citizen                           0                    0   
partner                                Yes                   No   
dependents                             Yes                   No   
tenure                                   9                    9   
phone_service                          Yes                  Yes   
multiple_lines                          No                  Yes   
online_security                         No                   No   
online_backup                          Yes                   No   
device_protection                       No                   No   
tech_support                           Yes                   No   
streaming_tv                           Yes                   No   
streaming_movies                        No                  Yes   
paperless_billing                      Yes                   No   
monthly_charges                       65.6                 59.9   
total_charges                        593.3                542.4   
churn                                   No                   No   
internet_service_type                  DSL                  DSL   
contract_type                     One year       Month-to-month   
payment_type                  Mailed check         Mailed check   
signup_date            2021-04-21 18:07:34  2021-04-21 18:07:34   

                                      2                    3     \
gender                                Male                 Male   
senior_citizen                           0                    1   
partner                                 No                  Yes   
dependents                              No                   No   
tenure                                   4                   13   
phone_service                          Yes                  Yes   
multiple_lines                          No                   No   
online_security                         No                   No   
online_backup                           No                  Yes   
device_protection                      Yes                  Yes   
tech_support                            No                   No   
streaming_tv                            No                  Yes   
streaming_movies                        No                  Yes   
paperless_billing                      Yes                  Yes   
monthly_charges                       73.9                 98.0   
total_charges                       280.85              1237.85   
churn                                  Yes                  Yes   
internet_service_type          Fiber optic          Fiber optic   
contract_type               Month-to-month       Month-to-month   
payment_type              Electronic check     Electronic check   
signup_date            2021-09-21 18:07:34  2020-12-21 18:07:34   

                                      4                        5     \
gender                              Female                   Female   
senior_citizen                           1                        0   
partner                                Yes                       No   
dependents                              No                      Yes   
tenure                                   3                        9   
phone_service                          Yes                      Yes   
multiple_lines                          No                       No   
online_security                         No                       No   
online_backup                           No                       No   
device_protection                       No                       No   
tech_support                           Yes                      Yes   
streaming_tv                           Yes                      Yes   
streaming_movies                        No                      Yes   
paperless_billing                      Yes                      Yes   
monthly_charges        

# REPLACE
<div class="alert alert-warning">
    
* churn_month has nulls; replace nulls with 'active' DROP (ALL DATES DUPLICATED).
* multiple_lines has 3 unique values. REPLACE 'No phone service' with 'No'
* online_security has 3 unique values. REPLACE 'No internet service' with 'No'
* online_backup has 3 unique values. REPLACE 'No internet service' with 'No'
* device_protection has 3 unique values. REPLACE 'No internet service' with 'No'
* tech_support has 3 unique values. REPLACE 'No internet service' with 'No'
* streaming_tv has 3 unique values. REPLACE 'No internet service' with 'No'
* streaming_movies has 3 unique values. REPLACE 'No internet service' with 'No'
    
</div>

In [23]:
df.multiple_lines = df.multiple_lines.replace('No phone service', 'No')

df.online_security = df.online_security.replace('No internet service', 'No')
df.online_backup = df.online_backup.replace('No internet service', 'No')
df.device_protection = df.device_protection.replace('No internet service', 'No')
df.tech_support = df.tech_support.replace('No internet service', 'No')
df.streaming_tv = df.streaming_tv.replace('No internet service', 'No')
df.streaming_movies = df.streaming_movies.replace('No internet service', 'No')

In [24]:
df.multiple_lines.unique() 

array(['No', 'Yes'], dtype=object)

In [25]:
df.online_security.unique()  

array(['No', 'Yes'], dtype=object)

In [26]:
df.online_backup.unique()  

array(['Yes', 'No'], dtype=object)

In [27]:
df.device_protection.unique()

array(['No', 'Yes'], dtype=object)

In [28]:
df.tech_support.unique()

array(['Yes', 'No'], dtype=object)

In [29]:
df.streaming_tv.unique()

array(['Yes', 'No'], dtype=object)

In [30]:
df.streaming_movies.unique()

array(['No', 'Yes'], dtype=object)

# ENCODE.MAP
<div class="alert alert-warning">
    
* gender has 2 unique values.
* partner has 2 unique values.
* dependents has 2 unique values.
* phone_service has 2 unique values.
* paperless_billing has 2 unique values.
* churn has 2 unique values.
* multiple_lines will have 2 unique values. REPLACED 'No phone service' with 'No'
* online_security will have 2 unique values. REPLACED 'No internet service' with 'No'
* online_backup will have 2 unique values. REPLACED 'No internet service' with 'No'
* device_protection will have 2 unique values. REPLACED 'No internet service' with 'No'
* tech_support will have 2 unique values. REPLACED 'No internet service' with 'No'
* streaming_tv will have 2 unique values. REPLACED 'No internet service' with 'No'
* streaming_movies will have 2 unique values. REPLACED 'No internet service' with 'No'
    
</div>

In [31]:
df.gender.unique()

array(['Female', 'Male'], dtype=object)

In [32]:
df.partner.unique()

array(['Yes', 'No'], dtype=object)

In [33]:
df.dependents.unique()

array(['Yes', 'No'], dtype=object)

In [34]:
df.phone_service.unique()

array(['Yes', 'No'], dtype=object)

In [35]:
df.paperless_billing.unique()

array(['Yes', 'No'], dtype=object)

In [36]:
df.churn.unique()

array(['No', 'Yes'], dtype=object)

In [37]:
df.senior_citizen.unique()

array([0, 1])

In [38]:
df.multiple_lines.unique()

array(['No', 'Yes'], dtype=object)

In [39]:
df.online_security.unique()

array(['No', 'Yes'], dtype=object)

In [40]:
df.online_backup.unique()

array(['Yes', 'No'], dtype=object)

In [41]:
df.device_protection.unique()

array(['No', 'Yes'], dtype=object)

In [42]:
df.tech_support.unique()

array(['Yes', 'No'], dtype=object)

In [43]:
df.streaming_tv.unique()

array(['Yes', 'No'], dtype=object)

In [44]:
df.streaming_movies.unique()

array(['No', 'Yes'], dtype=object)

In [45]:
encoded_df = pd.DataFrame()
encoded_df['gender_encoded'] = df.gender.map({'Male': 1, 'Female': 0})
encoded_df['partner_encoded'] = df.partner.map({'Yes': 1, 'No': 0})
encoded_df['dependents_encoded'] = df.dependents.map({'Yes': 1, 'No': 0})
encoded_df['phone_service_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['paperless_billing_encoded'] = df.paperless_billing.map({'Yes': 1, 'No': 0})
encoded_df['churn_encoded'] = df.churn.map({'Yes': 1, 'No': 0})
encoded_df['multiple_lines_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['online_security_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['online_backup_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['device_protection_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['tech_support_encoded'] = df.tech_support.map({'Yes': 1, 'No': 0})
encoded_df['streaming_tv_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df['streaming_movies_encoded'] = df.phone_service.map({'Yes': 1, 'No': 0})
encoded_df.T

0     1     2     3     4     5     6     7     \
gender_encoded                0     1     1     1     0     0     0     1   
partner_encoded               1     0     0     1     1     0     1     1   
dependents_encoded            1     0     0     0     0     1     0     0   
phone_service_encoded         1     1     1     1     1     1     1     1   
paperless_billing_encoded     1     0     1     1     1     1     1     1   
churn_encoded                 0     0     1     1     1     0     0     0   
multiple_lines_encoded        1     1     1     1     1     1     1     1   
online_security_encoded       1     1     1     1     1     1     1     1   
online_backup_encoded         1     1     1     1     1     1     1     1   
device_protection_encoded     1     1     1     1     1     1     1     1   
tech_support_encoded          0     0     1     1     1     0     0     0   
streaming_tv_encoded          1     1     1     1     1     1     1     1   
streaming_movies_encoded      1     1     1     1     1     1     1     1   

                           8     9     ...  7033  7034  7035  7036  7037  \
gender_encoded                0     0  ...     0     1     0     1     1   
partner_encoded               0     1  ...     1     0     0     1     0   
dependents_encoded            0     1  ...     0     0     0     1     0   
phone_service_encoded         1     1  ...     1     1     1     1     1   
paperless_billing_encoded     1     1  ...     1     1     0     0     0   
churn_encoded                 0     0  ...     0     1     0     0     0   
multiple_lines_encoded        1     1  ...     1     1     1     1     1   
online_security_encoded       1     1  ...     1     1     1     1     1   
online_backup_encoded         1     1  ...     1     1     1     1     1   
device_protection_encoded     1     1  ...     1     1     1     1     1   
tech_support_encoded          0     0  ...     0     1     0     0     0   
streaming_tv_encoded          1     1  ...     1     1     1     1     1   
streaming_movies_encoded      1     1  ...     1     1     1     1     1   

                           7038  7039  7040  7041  7042  
gender_encoded                0     0     0     1     1  
partner_encoded               0     0     0     0     1  
dependents_encoded            0     0     0     0     1  
phone_service_encoded         1     1     1     1     0  
paperless_billing_encoded     1     1     0     1     0  
churn_encoded                 0     1     1     0     0  
multiple_lines_encoded        1     1     1     1     0  
online_security_encoded       1     1     1     1     0  
online_backup_encoded         1     1     1     1     0  
device_protection_encoded     1     1     1     1     0  
tech_support_encoded          0     1     1     0     0  
streaming_tv_encoded          1     1     1     1     0  
streaming_movies_encoded      1     1     1     1     0  

[13 rows x 7043 columns]

In [47]:
ee.nunique_column_all(df)

Male      3555
Female    3488
Name: gender, dtype: int64

0    5901
1    1142
Name: senior_citizen, dtype: int64

No     3641
Yes    3402
Name: partner, dtype: int64

No     4933
Yes    2110
Name: dependents, dtype: int64

1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: tenure, Length: 73, dtype: int64

Yes    6361
No      682
Name: phone_service, dtype: int64

No     4072
Yes    2971
Name: multiple_lines, dtype: int64

No     5024
Yes    2019
Name: online_security, dtype: int64

No     4614
Yes    2429
Name: online_backup, dtype: int64

No     4621
Yes    2422
Name: device_protection, dtype: int64

No     4999
Yes    2044
Name: tech_support, dtype: int64

No     4336
Yes    2707
Name: streaming_tv, dtype: int64

No     4311
Yes    2732
Name: streaming_movies, dtype: int64

Yes    4171
No     2872
Name: paperless_billing, dtype: int64

20.05    61
19.85    45
19.95    44
19.90    44
20.00    43
         ..
97.45     1


No     4999
Yes    2044
Name: tech_support, dtype: int64

No     5174
Yes    1869
Name: churn, dtype: int64

In [48]:
ee.nunique_column_all(encoded_df)

1    3555
0    3488
Name: gender_encoded, dtype: int64

0    3641
1    3402
Name: partner_encoded, dtype: int64

0    4933
1    2110
Name: dependents_encoded, dtype: int64

1    6361
0     682
Name: phone_service_encoded, dtype: int64

1    4171
0    2872
Name: paperless_billing_encoded, dtype: int64

0    5174
1    1869
Name: churn_encoded, dtype: int64

1    6361
0     682
Name: multiple_lines_encoded, dtype: int64

1    6361
0     682
Name: online_security_encoded, dtype: int64

1    6361
0     682
Name: online_backup_encoded, dtype: int64

1    6361
0     682
Name: device_protection_encoded, dtype: int64

0    5174
1    1869
Name: tech_support_encoded, dtype: int64

1    6361
0     682
Name: streaming_tv_encoded, dtype: int64

1    6361
0     682
Name: streaming_movies_encoded, dtype: int64



0    5174
1    1869
Name: tech_support_encoded, dtype: int64

0    5174
1    1869
Name: churn_encoded, dtype: int64

# PIVOT
<div class="alert alert-warning">
    
* internet_service_type has 3 unique values.
* contract_type has 3 unique values.
* payment_type has 4 unique values.

</div>

In [ ]:
# Use pandas dummies to pivot features with more than two string values
# into multiple columns with binary int values that can be read as boolean
# drop_first = False in draft for human readability; Final will have it set to True.

dummy_df = pd.get_dummies(data=df[['internet_service_type', 'contract_type', 'payment_type']], drop_first=False)
dummy_df.T

In [ ]:
ee.nunique_column_all(dummy_df)

In [ ]:
df = pd.concat([df, encoded_df, dummy_df], axis=1)

In [ ]:
df.T

### DROP
<div class="alert alert-warning">
</div>

In [ ]:
# Drop encoded.map columns
df = df.drop(columns=['gender',
                      'partner',
                      'dependents',
                      'phone_service',
                      'paperless_billing',
                      'churn',
                      'multiple_lines',
                      'online_security',
                      'online_backup',
                      'device_protection',
                      'tech_support',
                      'streaming_tv',
                      'streaming_movies'])

# Drop pivot columns
df = df.drop(columns=['internet_service_type', 
                      'contract_type', 
                      'payment_type'])

In [ ]:
ee.nunique_column_all(df)

In [ ]:
df.T

In [ ]:
df.info()

In [ ]:
df.total_charges.info()

In [ ]:
df.total_charges = df.total_charges.replace(' ', '0')

In [ ]:
#df['total_charges'] = df['total_charges'].astype(float)
#df.total_charges = df.total_charges.astype(float)
df['total_charges'] = pd.to_numeric(df['total_charges'])

In [ ]:
df.total_charges.info()

In [ ]:
df.info()

In [ ]:
df["signup_date"] = pd.to_datetime(df["signup_date"])

In [ ]:
df.signup_date.info()

In [ ]:
df.info()

# Data Dictionary
<div class="alert alert-success">

| Feature | Definition |
|:--------|:-----------|
|Rated| True or False, The game's result is reflected in each player's rating|
|Winning Pieces| The color of pieces the winning player was moving|
|White Rating| Rating of the player moving the white pieces using the Glicko-2 rating method for games played on Lichess|
|Black Rating| Rating of the player moving the white pieces using the Glicko-2 rating method for games played on Lichess|
|Rating Difference| The difference in rating between the players in the game|
|Game Rating| The average rating of the two players in the game|
|Lower Rated White| True or False, The lower rated player is moving the white pieces|
|Opening Name| The name of the opening played in the game|
|Time Control Group| The amount of time allotted to each player to make their moves, **Standard** (60 min or more), **Rapid** (30 - 15 min), **Blitz** (5 - 3 min), or **Bullet** (2 or less), **Other** (any other time limit)|
|Upset (Target)| True or False, The lower rated player won the game|
|Additional Features|Encoded and values for categorical data and scaled versions continuous data|

</div>

In [ ]:
df.columns

# SPLIT  
<div class="alert alert-success">
    
    
    
</div>

In [ ]:
# use predefined function to split data for train, validate, and test
target = 'churn_encoded'
train_df, validate_df, test_df = w.split(df, target)

# VISUALIZATION
<div class="alert alert-sucess">
    
    
    
</div>

In [ ]:
train_df.T

In [ ]:
float_cols = train_df.select_dtypes(include='float').columns
float_cols

In [ ]:
# Plot numeric columns
for col in float_cols:
    plt.hist(train_df[col])
    plt.title(col)
    plt.show()
    plt.boxplot(train_df[col])
    plt.title(col)
    plt.show()

In [ ]:
train_df.churn_encoded.value_counts()

In [ ]:
plt.hist(df.churn_encoded)

In [ ]:
plt.hist(train_df.signup_date)

In [ ]:
churn = train_df[train_df['churn_encoded'] == 1]
active = train_df[train_df['churn_encoded'] == 0]

In [ ]:
train_df.columns

In [ ]:
# count plot on single categorical variable
sns.countplot(x ='tenure', data = train_df)
 
# Show the plot
plt.show()

In [ ]:
# count plot on two categorical variable
sns.countplot(x ='tenure', hue = "churn_encoded", data = train_df)#, palette = "Set2")
 
# Show the plot
plt.show()

In [ ]:
# Plotting the pie chart for above dataframe
train_df.groupby(['churn_encoded']).sum().plot(kind='pie', y='tenure')

* There appears to be a significant relationship that those with more tenure are less likely to churn.

In [ ]:
tech_churn = pd.pivot_table(train_df, 
                        index=[], 
                          columns=['tech_support_encoded'], 
                          values=['churn_encoded'], 
                          aggfunc='count')
tech_churn 

In [ ]:
plt.figure()
tech_churn.plot.bar(title='Tech Churn', stacked=True)
plt.xlabel('Customers')
plt.ylabel('Churn') 

In [ ]:
# count plot on two categorical variable
sns.countplot(x ='tech_support_encoded', hue = "churn_encoded", data = train_df)#, palette = "Set2")
 
# Show the plot
plt.show()

* Those with Tech Support Churn less than those without

In [ ]:
train_df.columns

In [ ]:
# count plot along y axis
sns.countplot(y ='tech_support_encoded', hue = "churn_encoded", data = train_df)
 
# Show the plot
plt.show()

* Those with Tech Support appear to be approximately 1/3 less likely to churn

In [ ]:
combos = [
    ('monthly_charges', 'total_charges')
#    ('internet_service_type_DSL', 'internet_service_type_Fiber optic')
#    ('sepal_length', 'petal_length'),
#    ('sepal_width', 'petal_width'),
#    ('sepal_length', 'petal_width'),
#    ('sepal_width', 'petal_length')
]

In [ ]:
x = combos[0][0]
y = combos[0][1]
def scatter_churn(x, y):
    for churn_encoded, subset in train_df.groupby('churn_encoded'):
        plt.scatter(subset[x], subset[y], label=churn_encoded)
        plt.xlabel(x)
        plt.ylabel(y)
        plt.legend()

In [ ]:
scatter_churn(x, y)

In [ ]:
plt.figure()
seniors.plot.bar(title='Total Churn – Seniors)', color='darkorange'); plt.xlabel('Age Groups'); plt.ylabel('Total Calls') 

# ANALYSIS
<div class="alert alert-success">
* TARGET = CHURN

## What do I want to ask?
    
### 1. Is there a relationship between Tenure and Churn?
* ```Tenure``` has a significant ```relationship``` with ```Churn```.

### 1.1 Is the average Tenure of Active customers greater than the average tenure of churned customers?
* The ```average Tenure``` of ```Active``` customers is ```less than or equal``` to the ```average Tenure``` of customers that ```Churn```.
    
### 2. Is there a relationship between Monthly Charges and Churn?  
* ```Monthly Charges``` has a significant ```Relationship``` with ```Churn```.  
    
### 2.1 Are the average monthly charges of customers that Churn higher than the average monthly charges of Active customers?
* The ```average monthly charges``` of customers that ```Churn``` are ```less than or equal``` to the ```average monthly charges``` of ```Active``` customers.

### 3. Is there a relationship between tech support and Churn?**  
* ```Tech Support``` has **NO** significant ```Relationship``` with ```Churn```**
    
### 3.1 Do customers with tech support churn less than customers without tech support?**  
* answer

</div>

In [ ]:
active = train_df.churn_encoded == 0
churn = train_df.churn_encoded == 1

<div class="alert alert-info">


**POSITIVE (+)** = insert Positive statement here  
**NEGATIVE (-)** = insert Negative statement here  

**RECALL**  
TP / (TP + FN)  
Use for less **Type II** errors when **FN** is worst outcome  
Maximize for **RECALL** if Cost of **FN** > Cost of **FP**  

**PRECISION**  
TP / (TP + FP)  
Use for less **Type I** errors when **FP** is worst outcome  
Maximize for **PRECISION** if Cost of **FP** > Cost of **FN**  

**ACCURACY**  
(TP + TN)/(FP+FN+TP+TN)  
total # prediction TRUE / total  
Maximize for **ACCURACY** if neither **RECALL** or **PRECISION** outweigh eachother  

* **Classification Confusion Matrix** (actual_col, prediction_row)(Positive_first, Negative_second)
|                     | actual Positive (+) | actual Negative(-) |
|---------------------|---------------------|--------------------|
|  pred Positive (+)  |     TP              |     FP (Type I)    |
|  pred Negative (-)  |     FN (Type II)    |     TN             |  

* <b>sklearn Confusion Matrix</b> (prediction_col, actual_row)(Negative_first, Positive_second)
|                     | pred Negative(-) | pred Positive (+) |
|---------------------|------------------|-------------------|
| actual Negative (-) |        TN        |    FP (Type I)    |
| actual Positive (+) |   FN (Type II)   |         TP        |  

**FP**: We **predicted** it was a **POSITIVE** when it was **actually** a **NEGATIVE**  
*    FP = We **FALSE**LY predicted it was **POSITIVE**  
* False = Our prediction was False, it was actually the opposite of our prediction  
* Oops... **TYPE I** error!  

**FN**: We **predicted** it was a **NEGATIVE** when it was **actually** a **POSITIVE**  
*    FN = We **FALSE**LY predicted it was **NEGATIVE**  
* False = Our prediction was False, it was actually the opposite of our prediction  
* Oops... **TYPE II** error!  

**TP**: We **predicted** it was a **POSITIVE** and it was **actually** a **POSITIVE**  
*   TP = We **TRUE**LY predicted it was **POSITIVE**  
* True = Our prediction was True, it was actually the same as our prediction  

**TN**: We **predicted** it was a **NEGATIVE** and it was **actually** a **NEGATIVE**  
*   TN = We **TRUE**LY predicted it was **NEGATIVE**  
* True = Our prediction was True, it was actually the same as our prediction  

</div>

## 1. Is there a relationship between Tenure and Churn
* ```Tenure``` has a significant ```relationship``` with ```Churn```.


**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **two_tail (tenure, churn_encoded)**  


* One Sample or Two Samples?\
 **two_sample (tenure, churn_encoded)**  


* Continuous or Discreat?\
 **Discreat (churn_encoded) vs Continuous (tenure) = T-Test**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\.  
 **$H_{0}$: Tenure has **NO** ```relationship``` with Churn**\.   
```train_df.tenure ``` != ```train_df.churn_encoded```  
 
 
* $𝐻_𝑎$: What am I trying to prove\.  
 **$H_{a}$: Tenure has a ```relationship``` with Churn**\.   
```train_df.tenure ``` == ```train_df.churn_encoded```

**Verify Assumptions**
* How big is sample size?
* If sample size is significantly > 30, we don't have to worry about normality (CLT!!)

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.tenure.count(), train_df.churn_encoded.count()

# T-Test
**Compute test statistic and probability (t-statistic & p-value)**

In [ ]:
t_stat, p_val = stats.levene(train_df.tenure, train_df.churn_encoded)

if p_val < α:
    print('equal_var = False (we cannot assume equal variance)')
    #Using Scipy 
    t_stat, p_val = stats.ttest_ind(train_df.tenure, train_df.churn_encoded, equal_var = False)
    print(f't-stat: {t_stat}')
    print(f'p-value: {p_val}')
else:
    print('equal_var = True (we will assume equal variance)')
    #Using Scipy 
    t_stat, p_val = stats.ttest_ind(train_df.tenure, train_df.churn_encoded, equal_var = True)
    print(f't-stat: {t_stat}')
    print(f'p-value: {p_val}')

In [ ]:
t_stat, p_val = stats.ttest_ind(train_df.tenure,
                                train_df.churn_encoded, 
                                equal_var=False)
t_stat, p_val

In [ ]:
"""
# One-Tail
if (t > 0) and ((p / 2) < α):
    print('We can reject the null hypothesis')
else:
    print('we fail to reject the null hypothesis')
"""

# Two-Tail
if p_val < α:
    print('We can reject the null hypothesis')
else:
    print('we fail to reject the null hypothesis')

Tenure has a ```relationship``` with Churn

## 1.1 Is the average Tenure of Active customers greater than the average tenure of churned customers?
* The ```average Tenure``` of ```Active``` customers is ```less than or equal``` to the ```average Tenure``` of customers that ```Churn```.

**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **one_tail (active_tenure.mean > churn_tenure.mean)**  


* One Sample or Two Samples?\
 **two_sample (active_tenure.mean > churn_tenure.mean)**  


* Continuous or Discreat?\
 **Continuous (active_tenure) vs Continuous (churn_tenure) = Pearson's R**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\. 
**$H_{0}$: Tenure of Active customers is ```<=``` Tenure of customers that Churn**   
```train_df.active_tenure.mean``` <= ```train_df.churn_tenure.mean```  


* $𝐻_𝑎$: What am I trying to prove\. 
**$H_{a}$: Tenure of Active customers is ```>``` Tenure of customers that Churn**  
```train_df.active_tenure.mean ``` > ```train_df.churn_tenure.mean```

**Verify Assumptions**
* How big is sample size?
* If sample size is significantly > 30, we don't have to worry about normality (CLT!!)

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.groupby(active).tenure.count(), 
train_df.groupby(churn).tenure.count()

In [ ]:
train_df.tenure.count()

# Pearson's R
**Compute test statistic and probability (r & p_value)**

In [ ]:
r, p_val = stats.pearsonr(train_df.groupby(active).tenure.mean(),
                      train_df.groupby(churn).tenure.mean())
r, p_val

In [ ]:
if p_val < α:
    print('Reject the null hypothesis')
else:
    print('Fail to reject the null hypothesis')

The average Tenure of Active customers is less than or equal to the average Tenure of customers that Churn

## 2. Is there a relationship between monthly_charges and Churn?
* ```Monthly Charges``` has a significant ```Relationship``` with ```Churn```.  

**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **two_tail (monthly_charges, churn_encoded)**  


* One Sample or Two Samples?\
 **two_sample (monthly_charges, churn_encoded)**  


* Continuous or Discreat?\
 **Discreat (churn_encoded) vs Continuous (monthly_charges) = T-Test**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\. 
**$H_{0}$: Monthly Charges does **NOT** have a significant ```Relationship``` with Churn**\.   
```train_df.monthly_charges.mean``` != ```train_df.churn_encoded.mean```  


* $𝐻_𝑎$: What am I trying to prove\. 
**$H_{a}$: Monthly Charges has a significant ```Relationship``` with Churn**  
```train_df.monthly_charges.mean``` == ```train_df.churn_encoded.mean```

**Verify Assumptions**
* How big is sample size?
* If sample size is significantly > 30, we don't have to worry about normality (CLT!!)

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.monthly_charges.count(), train_df.churn_encoded.count()

# T-Test
**Compute test statistic and probability (t-statistic & p-value)**

In [ ]:
t_stat, p_val = stats.levene(train_df.monthly_charges, train_df.churn_encoded)

if p_val < α:
    print('equal_var = False (we cannot assume equal variance)')
    #Using Scipy 
    t_stat, p_val = stats.ttest_ind(train_df.monthly_charges, 
                                    train_df.churn_encoded, 
                                    equal_var = False)
    print(f't-stat: {t_stat}')
    print(f'p-value: {p_val}')
else:
    print('equal_var = True (we will assume equal variance)')
    #Using Scipy 
    t_stat, p_val = stats.ttest_ind(train_df.monthly_charges, 
                                    train_df.churn_encoded, 
                                    equal_var = True)
    print(f't-stat: {t_stat}')
    print(f'p-value: {p_val}')

In [ ]:
t_stat, p_val = stats.ttest_ind(train_df.monthly_charges, 
                                train_df.churn_encoded, 
                                equal_var=False)
t_stat, p_val

In [ ]:
"""
# One-Tail
if (t > 0) and ((p / 2) < α):
    print('We can reject the null hypothesis')
else:
    print('we fail to reject the null hypothesis')
"""

# Two-Tail
if p_val < α:
    print('We can reject the null hypothesis')
else:
    print('we fail to reject the null hypothesis')

Monthly Charges has a significant ```Relationship``` with Churn

## 2.1 Are the average monthly charges of customers that Churn higher than the average monthly charges of Active customers?
* Average monthly charges of customers that Churn are ```less than or equal``` to the average monthly charges of Active customers

**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **two_tail (churn_monthly_charges > active_monthly_charges)**  


* One Sample or Two Samples?\
 **two_sample (churn_monthly_charges > active_monthly_charges)**  


* Continuous or Discreat?\
 **Continuous (churn_monthly_charges) vs Continuous (active_monthly_charges) = Pearson's R**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\. 
**$H_{0}$: Average monthly charges of customers that Churn are ```less than or equal``` to the average monthly charges of Active customers**   
```train_df.churn_monthly_charges.mean ``` <= ```train_df.active_monthly_charges.mean```  


* $𝐻_𝑎$: What am I trying to prove\. 
**$H_{a}$: Average monthly charges of customers that Churn are ```greater than``` the average monthly charges of Active customers**  
```train_df.churn_monthly_charges.mean ``` > ```train_df.active_monthly_charges.mean```

**Verify Assumptions**
* How big is sample size?
* If sample size is significantly > 30, we don't have to worry about normality (CLT!!)

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.groupby(churn).monthly_charges.count(), 
train_df.groupby(active).monthly_charges.count()

In [ ]:
train_df.monthly_charges.count()

# Pearson's R
**Compute test statistic and probability (r & p_val)**

In [ ]:
r, p_val = stats.pearsonr(train_df.groupby(churn).monthly_charges.mean(), 
                          train_df.groupby(active).monthly_charges.mean())
                          
r, p_val

In [ ]:
if p_val < α:
    print('Reject the null hypothesis')
else:
    print('Fail to reject the null hypothesis')

Monthly Charges of customers who Churn is <= Monthly Charges of Active customers

## 3. Is there a relationship between tech support and Churn?

**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **two_tail (tech_support_encoded, churn_encoded)**  


* One Sample or Two Samples?\
 **two_sample (tech_support_encoded, churn_encoded)**  


* Continuous or Discreat?\
 **Discreat (tech_support_encoded) vs Discreat (churn_encoded) = $Chi^2$**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\. 
**$H_{0}$: Tech Support has **NO** significant ```Relationship``` with Churn**\.    
```train_df.tech_support_encoded.mean ``` != ```train_df.churn_encoded.mean```  


* $𝐻_𝑎$: What am I trying to prove\. 
**$H_{a}$: Tech Support has a significant ```Relationship``` with Churn**  
```train_df.tech_support_encoded.mean ``` == ```train_df.churn_encoded.mean```

**Verify Assumptions**
* How big is sample size?
* If sample size is significantly > 30, we don't have to worry about normality (CLT!!)

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.tech_support_encoded.count(), train_df.churn_encoded.count()

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.groupby(train_df.tech_support_encoded==1).mean(), train_df.groupby(train_df.churn_encoded==1).mean()

In [ ]:
ee.nunique_column_all(train_df)

0    3104
1    1121
Name: churn_encoded, dtype: int64

0    3104
1    1121
Name: tech_support_encoded, dtype: int64

# $Chi^2$
**Compute test statistic and probability (Observed, Expected, degf, p-value)**

In [ ]:
tech_array = np.dot(train_df.groupby(train_df.tech_support_encoded==1).mean(),
              train_df.groupby(train_df.tech_support_encoded==1).mean().T)
tech_array

In [ ]:
churn_array = np.dot(train_df.groupby(train_df.churn_encoded==1).mean(),
              train_df.groupby(train_df.churn_encoded==1).mean().T)
churn_array

In [ ]:
observed = pd.crosstab(tech_array, churn_array)
observed

In [ ]:
chi2, p_val, degf, expected = stats.chi2_contingency(observed)

In [ ]:
#output values
print('Observed')
print(observed.values)
print('\nExpected')
print(expected.astype(int))
print('\n----')
print(f'Chi^2 = {chi2:.4f}')
print(f'p_val = {p}')

In [ ]:
if p_val < α:
    print('We reject the null hypothesis')
else:
    print('We fail to reject the null hypothesis')

Tech Support has *NO** Relationship with Churn

## 3.1 Is the average of Active customers with tech support churn less than customers without tech support?

**Set Hypothesis**  

* One Tail (```<= | >```) or Two Tails (```== | !=```)?\
 **one_tail (active_tech_support.mean > churn_tech_support.mean)**  


* One Sample or Two Samples?\
 **two_sample (active_tech_support.mean > churn_tech_support.mean)**  


* Continuous or Discreat?\
 **Continuous (active_tech_support) vs Continuous (churn_tech_support) = Pearson's R**  
     * T-Test = ```Discreat``` vs ```Continuous```
     * Pearson’s = ```Continuous``` vs ```Continuous``` (linear)
     * $Chi^2$ = ```Discreat``` vs ```Discreat```


* $𝐻_0$: The opposite of what I am trying to prove\. 
**$H_{0}$: The ```average``` of ```Active``` customers with ```Tech Support``` is ```less than or equal``` to the ```average``` of customers with ```Tech Support``` that have ```Churn```**   
* ```train_df.active_tech_support.mean ``` <= ```train_df.churn_tech_support.mean```


* $𝐻_𝑎$: What am I trying to prove\. 
**$H_{a}$: Avg of Active customers with Tech Support is ```>``` Avg of Churn customers with Tech Support**  
* ```train_df.active_tech_support.mean ``` > ```train_df.churn_tech_support.mean```

In [ ]:
# how many observations do I have for each sample? (N>30,we we meet normality condition)

train_df.groupby(active).tech_support_encoded.count(), 
train_df.groupby(churn).tech_support_encoded.count()

In [ ]:
train_df.tech_support_encoded.count()

# Pearson's R
**Compute test statistic and probability (r & p_val)**

In [ ]:
r, p_val = stats.pearsonr(train_df.groupby(active).tech_support_encoded.mean(),
                          train_df.groupby(churn).tech_support_encoded.mean())
                          
r, p_val

In [ ]:
if p_val < α:
    print('Reject the null hypothesis')
else:
    print('Fail to reject the null hypothesis')

Avg of Active customers with Tech Support is <= Avg of Churn customers with Tech Support

In [ ]:
notech = train_df[train_df['tech_support_encoded']==0]
notech_churn = train_df[(train_df.tech_support_encoded==0) & (train_df.churn_encoded==1)]
notech_active = train_df[(train_df.tech_support_encoded==0) & (train_df.churn_encoded==0)]

In [ ]:
notech_churn.T

In [ ]:
notech = df[df['tech_support_encoded']==0]
notech_churn = df[(df.tech_support_encoded==0) & (df.churn_encoded==1)]
notech_active = df[(df.tech_support_encoded==0) & (df.churn_encoded==0)]

In [ ]:
notech_churn.T

# MODELING

<div class="alert alert-success">
</div>